csharp_dotnetcore/60.ConversationHistory/ConversationHistoryBot.cs

```diff
@@ -0,0 +1,89 @@
+        {
+            _transcriptStore = transcriptStore;
+        }
+
+        /// <summary>
+        /// Every conversation turn for our Echo Bot will call this method.
+        /// There are no dialogs used, since it's "single turn" processing, meaning a single
+        /// request and response.
+        /// </summary>
+        /// <param name="turnContext">A <see cref="ITurnContext"/> containing all the data needed
+        /// for processing this conversation turn. </param>
+        /// <param name="cancellationToken">(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A <see cref="Task"/> that represents the work queued to execute.</returns>
+        /// <seealso cref="BotStateSet"/>
+        /// <seealso cref="ConversationState"/>
+        /// <seealso cref="IMiddleware"/>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (turnContext.Activity.Type == ActivityTypes.Message)
+            {
+                // Echo back to the user whatever they typed.
+                var responseMessage = $"You sent '{turnContext.Activity.Text}'\n";
+                await turnContext.SendActivityAsync(responseMessage);
+            }
+
+            // Send over the transcript when a request arrives. This is something that the client and the Bot have
+            // to agree on.
+            var shouldUploadHistory = false;
+            if (shouldUploadHistory)
+            {
+                var connectorClient = turnContext.TurnState.Get<ConnectorClient>(typeof(IConnectorClient).FullName);
+                var activity = turnContext.Activity;
+
+                // Get all the message type activities from the Transcript.
+                string continuationToken = null;
+                var activities = new List<Activity>();
+                do
+                {
+                    var pagedTranscript = await _transcriptStore.Store.GetTranscriptActivitiesAsync(activity.ChannelId, activity.Conversation.Id);
+                    activities.AddRange(pagedTranscript.Items
+                                        .Where(a => a.Type == ActivityTypes.Message)
+                                        .Select(ia => (Activity)ia)
+                                        .ToList());
+                    continuationToken = pagedTranscript.ContinuationToken;
+                }
+                while (continuationToken != null);
+
+                // Construct a Transcript object from the activities above and use the
+                // SendConversationHistoryAsync API to upload the historic activities.
+                var transcript = new Bot.Schema.Transcript(activities);
+                await connectorClient.Conversations.SendConversationHistoryAsync(activity.Conversation.Id, transcript);
+            }
+        }
+    }
```

Comment:
I'm missing something important here. I don't understand the relationship between the transcriptStore and the need to sendConversationHistory. Can you please add in comments explaining the logical flow? 

I *also* don't understand the point of this sample. There's the "always false" flag, that if is set to true, will cause huge amounts of downloads on every single turn (aka: all messages from the Transcript store). 100% of that store will then be uploaded. 

On then next turn ,that'll happen again. 

The bot seems like it should do something else as the sample here seems like it's not-at-all a best practice.

Reply:
Fixed as discussed

Reply:
Added comments.

---
In reply to: [218235887](https://github.com/Microsoft/BotBuilder-Samples/pull/391#discussion_r218235887) [](ancestors = 218235887)